New module - meraki_static_route

kbreit opened the pull request:
##### SUMMARY
This module will allow CRUD functionality against static routes.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
meraki_static_route

kbreit committed:
Initial commit for static route module
- Create and query functionality in place today

kbreit committed:
Enabled querying both all routes and single route

kbreit committed:
Bug fixes and idempotency check
- Requires new idempotency code to validate, blocking on that

kbreit committed:
Enabled test cases

kbreit committed:
Documentation fixes

kbreit committed:
Whitespace fix

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,243 @@
+options:
+    auth_key:
+        description:
+        - Authentication key provided by the dashboard. Required if environmental variable MERAKI_KEY is not set.
```
Make this two lines. The first line is always a summary of the parameter.
Next options clarify its use, format, etc...
```suggestion
        - Authentication key provided by the dashboard.
        - Required if environment variable C(MERAKI_KEY) is not set.
      type: str
```

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,243 @@
+    state:
+        description:
+        - Create or modify an organization.
+        choices: [absent, present, query]
```
Spaces !
```suggestion
        choices: [ absent, present, query ]
```

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,243 @@
+        - Authentication key provided by the dashboard. Required if environmental variable MERAKI_KEY is not set.
+    state:
+        description:
+        - Create or modify an organization.
```
```suggestion
        - Create or modify an organization.
      type: str
```
Please add types for all parameters !

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,243 @@
+    state: query
+    org_name: YourOrg
+  delegate_to: localhost
+- name: Query network named MyNet in the YourOrg organization
```
Keep a blank line between examples for readability.

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,243 @@
+
+EXAMPLES = r'''
+- name: List all networks associated to the YourOrg organization
+  meraki_network:
```
Wrong module in examples ?

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,243 @@
+        name=dict(type='str'),
+        subnet=dict(type='str'),
+        gateway_ip=dict(type='str'),
+        state=dict(type='str', choices=['absent', 'query', 'present'], default='present'),
```
Sort alphabetically, add defaults before choices.
```suggestion
        state=dict(type='str', default='present', choices=['absent', 'present', 'query']),
```

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,243 @@
+    # args/params passed to the execution, as well as if the module
+    # supports check mode
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=False,
```
No check-mode ?

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,243 @@
+    meraki.url_catalog['delete'] = delete_urls
+
+    if not meraki.params['org_name'] and not meraki.params['org_id']:
+        meraki.fail_json(msg='org_name or org_id parameters are required')
```
We tend to quote names and values in messages for readability.
```suggestion
        meraki.fail_json(msg="Parameters 'org_name' or 'org_id' parameters are required")
```

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,243 @@
+    # want to make any changes to the environment, just return the current
+    # state with no modifications
+    if module.check_mode:
+        return meraki.result
```
This is wrong. It is too easy. Check-mode in this case does not bring anything to the table. The user might have made mistakes and it has not been tested as it would be for a real run.

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,243 @@
+        else:
+            meraki.result['data'] = get_static_routes(meraki, net_id)
+    elif meraki.params['state'] == 'present':
+        payload = dict()
```
Please create dicts immediately with the values you already have.

dagwieers reviewed (changes requested):
A bit disappointed, a lot of things you should have known ;-)

kbreit committed:
Major improvements based on Dag's commments
- Improved documentation and examples
- Fixed fixed IP assignments and reserved range parameters
- Improved integration tests
- Made check mode work

kbreit committed:
Rename str to string in documentation

kbreit committed:
Move back to str

kbreit committed:
Fix indentation

kbreit commented:
@dagwieers Please review when you can.

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,386 @@
+    net_name: YourNet
+    route_id: d6fa4821-1234-4dfa-af6b-ae8b16c20c39
+    fixed_ip_assignments:
+      - aa:bb:cc:dd:ee:ff:
```
I think this is weird and will confuse people. Why not simply go for a list of dictionaries, which is a lot easier than a list of dictionaries with dictionaries.
```yaml
    fixed_ip_assignments:
    - mac: aa:bb:cc:dd:ee:ff
      ip: 192.0.1.11
      comment: Server
```

Looking at the argspec, this is how it ought to be, even.

kbreit replied:
Fixed.

dagwieers commented on lib/ansible/modules/network/meraki/meraki_static_route.py:
```diff
@@ -0,0 +1,386 @@
+        subnet=dict(type='str'),
+        gateway_ip=dict(type='str'),
+        state=dict(type='str', default='present', choices=['absent', 'present', 'query']),
+        fixed_ip_assignments=dict(type='list', default=None, elements='dict', options=fixed_ip_arg_spec),
```
Strictly speaking the `default=None` is not needed.

kbreit replied:
Fixed

dagwieers reviewed (changes requested):
A few remarks

kbreit committed:
Fix documentation for example of fixed_ip_assignments
- Removed default=None for args

dagwieers reviewed (approved):
LGTM, well done
